File: BackEnd/script/puntos_idealesi.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

from socal_jaccard import socal_jaccard
from coorp3c import coorp3c

def puntos_idealesi(X, x):
    n, p = X.shape

    a = np.matmul(X, x)
    d = np.matmul((np.ones((n, p)) - X), (np.ones((1, p)) - x))
    s = (a + d) / p
    d = 2 * (np.ones((n, 1)) - s)
    st = np.transpose(s)
    dt = np.transpose(d)

    S_Sokal, S_Jaccard, D2_S, D2_J = socal_jaccard(X)
    Y, vaps, percent, acum = coorp3c(D2_S)
    B = np.matmul(Y, np.transpose(Y))
    b = np.diag(B)
    Lambda = np.diag(vaps[:p])
    y = 1/2 * np.matmul(np.linalg.inv(Lambda), np.matmul(np.transpose(Y), (b - d)))
    yt = np.transpose(y)

    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')
    ax.plot([yt[0]], [yt[1]], [yt[2]], '.r', markersize=12)

    ax.set_xlabel('Primera coordenada principal')
    ax.set_ylabel('Segunda coordenada principal')
    ax.set_zlabel('Tercera coordenada principal')
    ax.grid()

    plt.show()

    return yt

File: BackEnd/drive-download-20230605T003529Z-001/2 y 3 Programas Matlab/puntos_idealesi.m
%%codigo para incluir SOLAMENTE AL punto ideal
function [yt]=puntos_idealesi(X,x)
%X=[1 1 0 0 1 1; 1 1 1 0 0 1 ; 1 0 0 1 0 1;1 0 0 1 0 1;1 0 0 0 1 1 ;0 0 0 0 1 0 ];
%x=[1 1 0 0 0 1];
[n,p]=size(X);

%x=[1 0 1 1 0 1]; %de prueba
a=X*x'; d=(ones(n,p)-X)*(ones(1,p)-x)';
s=(a+d)/p;
d=2*(ones(n,1)-s);
st=s';
dt=d';

%despues de usar la funcion coorp
% Y es la matriz de coordenadas
[S_Sokal,S_Jaccard,D2_S,D2_J]=socal_jaccard(X);
%[Y,vaps,percent,acum]=coorp(D2_S);
[Y,vaps,percent,acum]=coorp3c(D2_S);
B=Y*Y';
b=diag(B);
[n,p]=size(Y);
Lambda=diag(vaps(1:p));
y=1/2*inv(Lambda)*Y'*(b-d);
yt=y';


%%-------
%%representacion de los datos
%%en dimension 3

%plot (y(1),y(2),'.r','MarkerSize',12);
plot3 (y(1),y(2),y(3),'.r','MarkerSize',12);
%plot3 (y(1),y(2),y(3),'.k','MarkerSize',12);
grid


